### DISCUSSION TITLE
start qudi without showing the manager

### DISCUSSION THREAD

**Discussion by kay-jahnke** [Ulm-IQO/qudi-core#10] [Q&A] [answered]
Is there a way to start qudi without the Manager GUI showing up?
If possible it would be good to control that from config and not from command line. Is that possible?

If I just have the window from a module, is there a way, that qudi automatically closes, when the last GUI has been closed?
Alternatively, could I listen for the close event of my GUI and then request qudi to shut down?

**Comment by Neverhorst** [comment, marked as answer]
1) Currently there is no built-in way to start qudi in "GUI mode" without the manager GUI. The original plan was to allow qudi addon packages to install their own "manager/main GUI" but this is not yet implemented. Feel free to do so. ;) A fairly easy fix could be introduced to allow a config option to toggle manager GUI startup. Only problem I see is how do you quit qudi then if not using a console?

2) This refers to my initial thoughts on introducing a mandatory `QMainWindow` attribute to `GuiBase`. Currently you can store a reference to the GUI module main window widget where ever you like. This does not allow for automatic window handling by the GUI base class. If the abc of a GUI module would enforce storing it in an abstract property or similar, something like you mentioned would be possible. I tested a manual workaround in some modules which does the job:
```python
class MyGuiModule(GuiBase):
    ...

    def on_activate(self):
        ...
        self._mw.close_action.triggered.connect(self._deactivate_self)
        ...

    def on_deactivate(self):
        ...
        self._mw.close_action.triggered.disconnect()
        self._mw = None
        ...

    def _deactivate_self(self):
        self._qudi_main.module_manager.deactivate_module(self._meta['name'])
```

**Reply by Neverhorst** [reply]
Forgot to mention an important detail:
In the workaround I described above, the GUI module will only deactivate itself upon triggering the `close_action`, a custom `QAction` used in the `QMainWindow` instance to provide e.g. "close" in a file menu. If you press the "X" in the corner of the Window itself (using the OS window manager and not the custom `QAction`) this will not trigger.
Therefore I also added these lines of code to the `QMainWindow` implementation:
```python
class MyMainWindow(PySide2.QtWidgets.QMainWindow):

    ...

    def closeEvent(self, event: QtGui.QCloseEvent) -> None:
        super().closeEvent(event)
        self.close_action.trigger()

    ...
```

**Comment by kay-jahnke** [comment]
Thanks for the tips.

1. I made a PR #14 for this option. We could include this into master.
2. Your comment lead me the right way. I actually did not want to only deactivate the gui module, I wanted to close qudi all-together.

Here is a code example for getting number 2 done:
```python
class MyMainWindow(PySide2.QtWidgets.QMainWindow):

    def __init__(*args. **kwargs):
        super().__init__(*args. **kwargs)
        self.close_action = QtWidgets.QAction('Close qudi')

    def closeEvent(self, event: QtGui.QCloseEvent) -> None:
        super().closeEvent(event)
        self.close_action.trigger()

    ...



class MyGuiModule(GuiBase):
    ...

    def on_activate(self):
        self._mw.close_action.triggered.connect(self._close_qudi)
        ...

    def on_deactivate(self):
        ...
        self._mw.close_action.triggered.disconnect(self._close_qudi)
        self._mw = None
        ...

    def _deactivate_self(self):
        self._qudi_main.quit()
```

The only disadvantage of this solution is, that qudi is (regularly) closed independent of what else was going on in the background. So if only one window is open, this is fine, but if you have more windows open, it will kill the other ones too.

As the regular `quit` is called, qudi shuts down properly and also saves status variables and logs.

This solution was tested and works fine on Win 10 Pro 21H2.

**Reply by Neverhorst** [reply]
I would rather implement a "quit qudi on last GUI module deactivated" option or similar.
But still... this should only come with an overhaul of the `GuiBase` class in order to avoid workarounds and make it a proper feature.
